bridge-ffi/src/lib.rs

```diff
@@ -0,0 +1,650 @@
+                    bridge::poll_workflow_activation_response::Error { message },
+                )),
+            };
+            unsafe {
+                callback(user_data, core.encode_proto(&resp).into_raw());
+            }
+            return;
+        }
+    };
+    let user_data = Box::new(UserDataHandle(user_data));
+    core.tokio_runtime.clone().spawn(async move {
+        let resp = bridge::PollWorkflowActivationResponse {
+            response: Some(match core.poll_workflow_activation(req).await {
+                Ok(act) => bridge::poll_workflow_activation_response::Response::Activation(act),
+                Err(err) => bridge::poll_workflow_activation_response::Response::Error(err),
+            }),
+        };
+        unsafe { callback(user_data.0, core.encode_proto(&resp).into_raw()) };
+    });
+}
+
+/// Poll activity task.
+///
+/// The req_proto and req_proto_len represent a byte array for a
+/// PollActivityTaskRequest protobuf message. The callback is invoked on
+/// completion with a PollActivityTaskResponse protobuf message.
+#[no_mangle]
+pub extern "C" fn tmprl_poll_activity_task(
+    core: *mut tmprl_core_t,
+    req_proto: *const u8,
+    req_proto_len: libc::size_t,
+    user_data: *mut libc::c_void,
+    callback: tmprl_callback,
+) {
+    let core = unsafe { &mut *core };
+    let req = match tmprl_core_t::decode_proto::<bridge::PollActivityTaskRequest>(
+        req_proto,
+        req_proto_len,
+    ) {
+        Ok(req) => req,
+        Err(message) => {
+            let resp = bridge::PollActivityTaskResponse {
+                response: Some(bridge::poll_activity_task_response::Response::Error(
+                    bridge::poll_activity_task_response::Error { message },
+                )),
+            };
+            unsafe {
+                callback(user_data, core.encode_proto(&resp).into_raw());
+            }
+            return;
+        }
+    };
+    let user_data = Box::new(UserDataHandle(user_data));
+    core.tokio_runtime.clone().spawn(async move {
+        let resp = bridge::PollActivityTaskResponse {
+            response: Some(match core.poll_activity_task(req).await {
+                Ok(task) => bridge::poll_activity_task_response::Response::Task(task),
+                Err(err) => bridge::poll_activity_task_response::Response::Error(err),
+            }),
+        };
+        unsafe { callback(user_data.0, core.encode_proto(&resp).into_raw()) };
+    });
+}
+
+impl tmprl_core_t {
+    async fn new(
+        tokio_runtime: std::sync::Arc<tokio::runtime::Runtime>,
+        req: bridge::InitRequest,
+    ) -> Result<tmprl_core_t, String> {
+        let mut core_opts = temporal_sdk_core::CoreInitOptionsBuilder::default();
+        if let Some(req_gateway_opts) = req.gateway_options {
+            let mut gateway_opts = temporal_sdk_core::ServerGatewayOptionsBuilder::default();
+            if !req_gateway_opts.target_url.is_empty() {
```

Comment:
All this conversion goop for all these structs should be moved into `TryInto` implementations.

I would also create a dedicated module/file for it.

Reply:
I didn't originally do `TryFrom` because I would have to either edit the proto project (I don't like adding functions to generated code if I can help it) or or the core one. But I agree I can move the conversions to their own module/file.

Reply:
I would add it to the proto project. It's not adding to generated code - the generated code is _included_ by that project, so I really don't think it's a problem. It's very reasonable to augment generated code with hand-written code, and if I hadn't done that for the existing protos things would be a huge mess. Being able to use the traits is very desirable as they compose well with other stuff.

Reply:
Ok, will do

Reply:
Could not add to proto project due to needing the core project to translate into. Added to bottom of this file.